### PR TITLE
Ultimate reflection pull request #2

### PR DESCRIPTION
**Pull request by xeno-by**
Supercedes https://github.com/scala/scala/pull/1048, https://github.com/scala/scala/pull/1052, https://github.com/scala/scala/pull/1054 and https://github.com/scala/scala/pull/1057.

### CODE REVIEW

**Comment by retronym** [comment]
@xeno-by if you hash-reference other pull requests, like #1048, #1052, #1054, #1057, the obedient robotic octocats will insert hrefs back and forth.

**Comment by xeno-by** [comment]
neat, thanks!

**Comment by xeno-by** [comment]
review by @odersky, @adriaanm and @paulp

**Comment by xeno-by** [comment]
PLS REBUILD ALL

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala [hunk 1]
```diff
@@ -152,13 +155,29 @@ trait JavaMirrors extends internal.SymbolTable with api.JavaUniverse { self: Sym
...
-      if (!owner.info.member(wannabe.name).alternatives.contains(wannabe)) ErrorNotMember(wannabe, owner)
+    private def checkMemberOf(wannabe: Symbol, owner: ClassSymbol) {
+      if (wannabe.owner == AnyClass || wannabe.owner == AnyRefClass || wannabe.owner == ObjectClass) {
+        // do nothing
+      } else if (wannabe.owner == AnyValClass) {
+        if (!owner.isPrimitiveValueClass && !owner.isDerivedValueClass) ErrorNotMember(wannabe, owner)
+      } else {
+        def isMemberOf(wannabe: Symbol, owner: ClassSymbol): Boolean = {
+          val isNonShadowedMember = owner.info.member(wannabe.name).alternatives.contains(wannabe)
+          isNonShadowedMember || owner.info.baseClasses.tail.exists(base => isMemberOf(wannabe, base.asClass))
```
Comment (odersky) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala]: can't you write:

```
 owner.info.baseClasses contains wannabe.owner
```

File: src/reflect/scala/reflect/api/Mirrors.scala [hunk 2]
```diff
@@ -226,7 +242,7 @@ trait Mirrors { self: Universe =>
      *  Such a mirror can be used to further reflect against the members of the object
      *  to get/set fields, invoke methods and inspect inner classes and objects.
      */
-    def reflect(obj: Any): InstanceMirror
```
Comment (odersky) [inline src/reflect/scala/reflect/api/Mirrors.scala:229]: Please add a comment about purpose of ClassTag

**Comment by odersky** [comment]
If the two inline comments are taken care of it will LGTM

**Comment by xeno-by** [comment]
PLS REBUILD ALL

**Pull request merged**